src/components/task-step/exercise.cjsx

```diff
@@ -52,48 +113,129 @@ module.exports = React.createClass
 
     if sectionsLinks.length > 0 then helpLink
 
+  stopListeningForProgress: (props) ->
+    props ?= @props
+    {taskId} = props
 
-  render: ->
-    {id, taskId} = @props
-    step = TaskStepStore.get(id)
-    task = TaskStore.get(taskId)
+    TaskProgressStore.off("update.#{taskId}", @setCurrentStepFromProgress)
+
+  startListeningForProgress: (props) ->
+    props ?= @props
+    {taskId} = props
+
+    TaskProgressStore.on("update.#{taskId}", @setCurrentStepFromProgress)
+
+  canAllContinue: ->
+    {parts} = @state
+
+    _.reduce parts, (previous, part) ->
+      previous and canOnlyContinue(part.id)
+    , true
+
```

Comment:
Not a big thing, but this and the `_.reduce` on line 153 below could use `_.every`.  that's a tiny bit more efficient since it'll abort the loop early as soon as something returns `false`.